components/brave_federated/learning_service.cc

```diff
@@ -0,0 +1,216 @@
+#include "base/time/time.h"
+#include "base/timer/timer.h"
+#include "brave/components/brave_federated/communication_adapter.h"
+#include "brave/components/brave_federated/config_utils.h"
+#include "brave/components/brave_federated/eligibility_service.h"
+#include "brave/components/brave_federated/features.h"
+#include "brave/components/brave_federated/resources/grit/brave_federated_resources.h"
+#include "brave/components/brave_federated/task/federated_task_runner.h"
+#include "brave/components/brave_federated/task/model.h"
+#include "brave/components/brave_federated/task/typing.h"
+#include "brave/components/brave_federated/util/synthetic_dataset.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+#include "ui/base/resource/resource_bundle.h"
+
+namespace brave_federated {
+
+namespace {
+
+absl::optional<TaskResult> LoadDatasetAndRunTask(
+    std::unique_ptr<FederatedTaskRunner> task_runner) {
+  auto synthetic_dataset = std::make_unique<SyntheticDataset>(500);
+  SyntheticDataset test_dataset = synthetic_dataset->SeparateTestData(50);
+
+  task_runner->SetTrainingData(synthetic_dataset->GetDataPoints());
+  task_runner->SetTestData(test_dataset.GetDataPoints());
+  VLOG(2) << "Model and data set. Task runner initialized.";
+
+  return task_runner->Run();
+}
+
+}  // namespace
+
+LearningService::LearningService(
+    EligibilityService* eligibility_service,
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory)
+    : url_loader_factory_(url_loader_factory),
+      eligibility_service_(eligibility_service) {
+  DCHECK(!init_task_timer_);
+
+  std::string data_resource;
+  auto& resource_bundle = ui::ResourceBundle::GetSharedInstance();
+  if (resource_bundle.IsGzipped(IDR_BRAVE_FEDERATED_CONFIG)) {
+    data_resource =
+        resource_bundle.LoadDataResourceString(IDR_BRAVE_FEDERATED_CONFIG);
+  } else {
+    data_resource = static_cast<std::string>(
+        resource_bundle.GetRawDataResource(IDR_BRAVE_FEDERATED_CONFIG));
+  }
+  DCHECK_GT(data_resource.size(), 0U);
+
+  lsc_ = std::make_unique<LearningServiceConfig>(data_resource);
+  const net::BackoffEntry::Policy reconnect_policy = lsc_->GetReconnectPolicy();
+  const net::BackoffEntry::Policy request_task_policy =
+      lsc_->GetRequestTaskPolicy();
+  const net::BackoffEntry::Policy post_results_policy =
+      lsc_->GetPostResultsPolicy();
+  model_spec_ = std::make_unique<ModelSpec>(lsc_->GetModelSpec());
+
+  communication_adapter_ = std::make_unique<CommunicationAdapter>(
+      url_loader_factory_, reconnect_policy, request_task_policy);
+  post_results_policy_ =
+      std::make_unique<const net::BackoffEntry::Policy>(post_results_policy);
+  post_results_backoff_entry_ =
+      std::make_unique<net::BackoffEntry>(post_results_policy_.get());
+
+  const int init_federated_service_wait_time_in_seconds =
```

Comment:
what are we waiting for?

Reply:
@atuchin-m suggested delaying the initialisation of this service since it's not critical and may slow down startup